File: AndroidWatchdog/app/src/main/java/com/viavr/watchdog/WatchdogConfig.java
package com.viavr.watchdog;

import android.util.Log;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.*;

public class WatchdogConfig {

    final String LOG_TAG = "WATCHDOG";

    public String launcherVersion;                                      // launcher_version
    public String watchdogTargetVersion;                                // watchdog_version
    public String targetPackage;                                        // target_package
    public String targetPackageActivity;                                // target_package_activity
    public String runtimeApiUrl;                                        // runtime_api_url
    public int checkProcessEveryMs = 5000;                              // check_every_milliseconds
    public boolean allowStartLauncher = false;                          // allow_start_vrlauncher
    public boolean allowReboot = false;                                 // allow_reboot
    public int rebootAtHour = 6;                                        // reboot_at_hour
    public boolean allowWakeup = false;                                 // allow_wakeup
    public int wakeupEverySeconds = 1800;                               // wakeup_every_seconds
    public int wakeupRandomSeconds = 600;                               // wakeup_random_seconds
    public boolean allowSendRuntime = false;                            // allow_runtime
    public int sendRuntimeEverySeconds = 1800;                          // runtime_every_seconds
    public int sendRuntimeRandomSeconds = 600;                          // runtime_random_seconds
    public boolean allowKillApps = true;                                // allow_kill_apps
    public ArrayList<String> launcherKillWhitelist = new ArrayList<>(); // launcher_kill_whitelist

    public HashMap<String, Long> packagesForegroundTimedWhitelist;
    public String errors;

    @Override
    public String toString() {
        return "[v" + BuildConfig.VERSION_NAME + "] received WatchdogConfig{" +
                "launcherVersion='" + launcherVersion + '\'' +
                ", watchdogTargetVersion='" + watchdogTargetVersion + '\'' +
                ", targetPackage='" + targetPackage + '\'' +
                ", targetPackageActivity='" + targetPackageActivity + '\'' +
                ", runtimeApiUrl='" + runtimeApiUrl + '\'' +
                ", checkProcessEveryMs=" + checkProcessEveryMs +
                ", allowStartLauncher=" + allowStartLauncher +
                ", allowReboot='" + allowReboot + '\'' +
                ", rebootAtHour='" + rebootAtHour + '\'' +
                ", allowWakeup='" + allowWakeup + '\'' +
                ", wakeup_every_seconds='" + wakeupEverySeconds + '\'' +
                ", wakeup_random_seconds='" + wakeupRandomSeconds + '\'' +
                ", allowSendRuntime='" + allowSendRuntime + '\'' +
                ", sendRuntimeEverySeconds='" + sendRuntimeEverySeconds + '\'' +
                ", sendRuntimeRandomSeconds='" + sendRuntimeRandomSeconds + '\'' +
                ", allow_kill_apps='" + allowKillApps + '\'' +
                ", launcher_kill_whitelist=" + (launcherKillWhitelist == null ? "null" : "[" + String.join(",", launcherKillWhitelist) + "]") +
                '}';
    }

    // парсинг конфига
    public static WatchdogConfig create(JSONObject jsonObject){

        WatchdogConfig watchdogConfig = new WatchdogConfig();

        watchdogConfig.errors = null;
        watchdogConfig.packagesForegroundTimedWhitelist = new HashMap<>();

        if(jsonObject == null) return watchdogConfig;

        String errorStart = "WatchdogConfig.create(): json error: ";

        try {
            watchdogConfig.launcherVersion = jsonObject.getString("launcher_version");
        } catch (JSONException e) {
            watchdogConfig.error(errorStart + e.getMessage());
        }

        try {
            watchdogConfig.watchdogTargetVersion = jsonObject.getString("watchdog_version");
        } catch (JSONException e) {
            watchdogConfig.error(errorStart + e.getMessage());
        }

        try {
            watchdogConfig.targetPackage = jsonObject.getString("target_package");
        } catch (JSONException e) {
            watchdogConfig.error(errorStart + e.getMessage());
        }

        try {
            watchdogConfig.targetPackageActivity = jsonObject.getString("target_package_activity");
        } catch (JSONException e) {
            watchdogConfig.error(errorStart + e.getMessage());;
        }

        try {
            watchdogConfig.runtimeApiUrl = jsonObject.getString("runtime_api_url");
        } catch (JSONException e) {
            watchdogConfig.error(errorStart + e.getMessage());;
        }

        try {
            watchdogConfig.checkProcessEveryMs = jsonObject.getInt("check_every_milliseconds");
        } catch (JSONException e) {
            watchdogConfig.error(errorStart + e.getMessage());
        }

        try {
            watchdogConfig.allowStartLauncher = jsonObject.getBoolean("allow_start_vrlauncher");
        } catch (JSONException e) {
            watchdogConfig.error(errorStart + e.getMessage());
        }

        try {
            watchdogConfig.allowReboot = jsonObject.getBoolean("allow_reboot");
        } catch (JSONException e) {
            watchdogConfig.error(errorStart + e.getMessage());
        }

        try {
            watchdogConfig.rebootAtHour = jsonObject.getInt("reboot_at_hour");
        } catch (JSONException e) {
            watchdogConfig.error(errorStart + e.getMessage());
        }

        try {
            watchdogConfig.allowWakeup = jsonObject.getBoolean("allow_wakeup");
        } catch (JSONException e) {
            watchdogConfig.error(errorStart + e.getMessage());
        }

        try {
            watchdogConfig.wakeupEverySeconds = jsonObject.getInt("wakeup_every_seconds");
        } catch (JSONException e) {
            watchdogConfig.error(errorStart + e.getMessage());
        }

        if(watchdogConfig.wakeupEverySeconds <= 0) watchdogConfig.allowWakeup = false;

        try {
            watchdogConfig.wakeupRandomSeconds = jsonObject.getInt("wakeup_random_seconds");
        } catch (JSONException e) {
            watchdogConfig.error(errorStart + e.getMessage());
        }

        try {
            watchdogConfig.allowSendRuntime = jsonObject.getBoolean("allow_runtime");
        } catch (JSONException e) {
            watchdogConfig.error(errorStart + e.getMessage());
        }

        try {
            watchdogConfig.sendRuntimeEverySeconds = jsonObject.getInt("runtime_every_seconds");
        } catch (JSONException e) {
            watchdogConfig.error(errorStart + e.getMessage());
        }

        if(watchdogConfig.sendRuntimeEverySeconds <= 0) watchdogConfig.allowSendRuntime = false;

        try {
            watchdogConfig.sendRuntimeRandomSeconds = jsonObject.getInt("runtime_random_seconds");
        } catch (JSONException e) {
            watchdogConfig.error(errorStart + e.getMessage());
        }

        try {
            watchdogConfig.allowKillApps = jsonObject.getBoolean("allow_kill_apps");
        } catch (JSONException e) {
            watchdogConfig.error(errorStart + e.getMessage());
        }

        try {
            JSONArray whitelist = jsonObject.getJSONArray("launcher_kill_whitelist");

            watchdogConfig.launcherKillWhitelist.clear();

            for(int i = 0; i < whitelist.length(); i++)
                watchdogConfig.launcherKillWhitelist.add(whitelist.getString(i));
        } catch (JSONException e) {
            watchdogConfig.error(errorStart + e.getMessage());
        }

        // если таргет приложение кривое
        if(watchdogConfig.targetPackage == null || watchdogConfig.targetPackageActivity == null){
            watchdogConfig.allowStartLauncher = false;
            watchdogConfig.error("WatchdogConfig.create(): watchdogConfig.allowStartLauncher set to FALSE, please check 'target_package' and 'target_package_activity'");
        }

        return watchdogConfig;
    }

    private void error(String message){
        if(errors == null) errors = "[" + message + "]";
        else errors += "[" + message + "]";

        Log.e(LOG_TAG, message);
    }

    public long getWhitelistEndTimestamp(String packageName){
        if(!packagesForegroundTimedWhitelist.containsKey(packageName)) return 0;

        return packagesForegroundTimedWhitelist.get(packageName);
    }

    public boolean checkIfWhitelisted(String packageName){

        if(!packagesForegroundTimedWhitelist.containsKey(packageName)) return false;

        if(System.currentTimeMillis() > packagesForegroundTimedWhitelist.get(packageName)){
            removeFromWhitelist(packageName);
            return false;
        }

        return true;
    }

    public void addToWhitelist(String packageName, Long accessTimeMilliseconds){
        packagesForegroundTimedWhitelist.put(packageName, System.currentTimeMillis() + accessTimeMilliseconds);
    }

    public void removeFromWhitelist(String packageName){
        packagesForegroundTimedWhitelist.remove(packageName);
    }

    public void clearWhitelist(){
        packagesForegroundTimedWhitelist.clear();
    }

    private static String[] jsonArrayToStringArray(JSONArray array) {
        if(array==null) return null;

        String[] arr = new String[array.length()];

        for(int i=0; i < arr.length; i++)
            arr[i] = array.optString(i);

        return arr;
    }
}

File: Assets/VIAVR/Scripts/WatchdogConfig.cs
using System;
using Newtonsoft.Json;
using UnityEngine;
using VIAVR.Scripts.Network;

namespace VIAVR.Scripts
{
    [CreateAssetMenu(fileName = "WatchdogConfigScriptable", menuName = "VIAVR/Application Configurations/Watchdog Config")]
    public class WatchdogConfig : ScriptableObject
    {
        [Serializable]
        public class JsonConfig
        {
            // package который надо чекать
            [SerializeField] [JsonProperty("target_package")] private string _targetPackage;
            // параметр для запуска _targetPackage
            [SerializeField] [JsonProperty("target_package_activity")] private string _targetPackageActivity;
        
            // чекать _targetPackage каждые _checkProcessEveryMs миллисекунд
            [SerializeField] [JsonProperty("check_every_milliseconds")] private int _checkProcessEveryMs;
            // можно ли запускать _targetPackage
            [SerializeField] [JsonProperty("allow_start_vrlauncher")] private bool _allowStartLauncher;
        
            // разрешить ребут через вачдог
            [SerializeField] [JsonProperty("allow_reboot")] private bool _allowReboot;
            // каждый день выполнять перезагрузку шлема в _everydayRebootAtHour часов (24-часовой формат) 00 минут
            [SerializeField] [JsonProperty("reboot_at_hour")] private int _everydayRebootAtHour;
        
            // разрешить вейкап через вачдог
            [SerializeField] [JsonProperty("allow_wakeup")] private bool _allowWakeup; // в данном проекте если _allowWakeup == false то через ватчдог пересылается аналитика в DeviceRuntimeUrl
            // просыпаться каждые _wakeupEverySeconds (+ рандом)
            [SerializeField] [JsonProperty("wakeup_every_seconds")] private int _wakeupEverySeconds;
            // добавлять рандомное количество секунд в промежутке 0 - _wakeupRandomSeconds к времени вейкапа
            [SerializeField] [JsonProperty("wakeup_random_seconds")] private int _wakeupRandomSeconds;
        
            // разрешить слать данные шлема через вачдог
            [SerializeField] [JsonProperty("allow_runtime")] private bool _allowSendRuntime;
            // слать данные шлема каждые _runtimeEverySeconds (+ рандом)
            [SerializeField] [JsonProperty("runtime_every_seconds")] private int _runtimeEverySeconds;
            // добавлять рандомное количество секунд в промежутке 0 - _runtimeRandomSeconds к времени вейкапа
            [SerializeField] [JsonProperty("runtime_random_seconds")] private int _runtimeRandomSeconds;
        
            // можно ли убивать процессы свёрнутых приложений (лаунчер тоже)
            [SerializeField] [JsonProperty("allow_kill_apps")] private bool _allowKillApps;
            // не убивать процесс лаунчера при запуске пакета приложения из списка
            [SerializeField] [JsonProperty("launcher_kill_whitelist")] private string[] _launcherKillWhitelist; //TODO автоматом подхватывать игры из GameCenter
            [JsonProperty("watchdog_version")] public string WatchdogTargetVersion { get; set; }
            [JsonProperty("launcher_version")] public string LauncherVersion { get; set; }
            [JsonProperty("runtime_api_url")] public string DeviceRuntimeUrl { get; set; } // используется если _allowWakeup == false
        }
    
        [SerializeField] private string _version;
        [SerializeField] private string _packageName;
        [SerializeField] private string _serviceClassName;
        [SerializeField] private string _reloadIntentAction;
        [SerializeField] private string _commandIntentAction;
        [SerializeField] private string _streamingAssetsApkRelativePath;

        [SerializeField] private JsonConfig _jsonConfig;
    
        public string Version => _version;
        public string PackageName => _packageName;
        public string ServiceClassName => _serviceClassName;
        public string ReloadIntentAction => _reloadIntentAction;
        public string CommandIntentAction => _commandIntentAction;
        public string StreamingAssetsApkRelativePath => _streamingAssetsApkRelativePath;
    
        public string SerializedJsonConfig
        {
            get
            {
                //_jsonConfig.SentryDSN = APIConnectionConfiguration.SENTRY_WATCHDOG_DSN;
                _jsonConfig.WatchdogTargetVersion = Version;
                _jsonConfig.LauncherVersion = Application.version;
                _jsonConfig.DeviceRuntimeUrl = APIConnectionConfiguration.DeviceRuntimeUrl;
            
                return JsonConvert.SerializeObject(_jsonConfig, Formatting.Indented);
            }
        }
    }
}